examples/hosts/app-integration/external-data/src/utilities/clientManager.ts

```diff
@@ -0,0 +1,122 @@
+	 * Values are the set of Fluid Container URLs that will be notified of changes.
+	 */
+	private readonly _taskListMapping: Map<ExternalTaskListId, Set<ClientSessionUrl>>;
+	/**
+	 * Map of active clients to external resource id.
+	 * Values are the set of external resource id's that the client has active and is registered to listen for.
+	 */
+	private readonly _clientMapping: Map<ClientSessionUrl, Set<ExternalTaskListId>>;
+
+	public constructor() {
+		this._clientMapping = new Map<ClientSessionUrl, Set<ExternalTaskListId>>();
+		this._taskListMapping = new Map<ExternalTaskListId, Set<ClientSessionUrl>>();
+	}
+
+	/**
+	 * Gets the current list of client session URLs for the specified task list id.
+	 */
+	public getClientSessions(externalTaskListId: ExternalTaskListId): Set<ClientSessionUrl> {
+		const clientSessionUrls = this._clientMapping.get(externalTaskListId);
+		if (clientSessionUrls === undefined) {
+			console.error(
+				`CUSTOMER SERVICE: "${externalTaskListId}" is not registered to a client session.`,
+			);
+		}
+		return clientSessionUrls ?? new Set<ClientSessionUrl>();
+	}
+
+	/**
+	 * Checks if the client and the external resource already have a mapping and returns the result.
+	 * The user can determine whether to re-register given this information.
+	 */
+	public needsNewMappingEntry(
+		client: ClientSessionUrl,
+		externalTaskListId: ExternalTaskListId,
+	): boolean {
+		let needsNewMappingEntry = true;
+		const clientTaskListIds = this._clientMapping.get(client);
+		// eslint-disable-next-line @typescript-eslint/prefer-optional-chain
+		if (clientTaskListIds !== undefined && clientTaskListIds.has(externalTaskListId)) {
+			const taskListClients = this._taskListMapping.get(client);
+			// eslint-disable-next-line @typescript-eslint/prefer-optional-chain
+			if (taskListClients !== undefined && taskListClients.has(client)) {
+				needsNewMappingEntry = false;
+			}
+		}
+		return needsNewMappingEntry;
+	}
+
+	/**
+	 * Returns a boolean if externalTaskListId already exists entry exists. This means that the customer service
+	 * is already subscribed for webhook notifications for it so we do not need to re-subscribe.
+	 */
+	public needsNewSubscription(externalTaskListId: ExternalTaskListId): boolean {
+		const externalTaskListIdClients = this._taskListMapping.get(externalTaskListId);
+		if (externalTaskListIdClients !== undefined) {
+			return externalTaskListIdClients.size > 0;
+		}
+		return true;
+	}
+
+	/**
+	 * Registers a client session url to an external resource id for the duration of the client session.
+	 */
+	public registerClient(client: ClientSessionUrl, externalTaskListId: ExternalTaskListId): void {
+		if (this._taskListMapping.get(client) === undefined) {
+			this._taskListMapping.set(client, new Set<ExternalTaskListId>([externalTaskListId]));
+		}
+
+		if (this._clientMapping.get(externalTaskListId) === undefined) {
+			this._clientMapping.set(externalTaskListId, new Set<ClientSessionUrl>([client]));
+		}
```

Comment:
Yeah I had them reversed as well, but then need to be this way for it to work! Potentially I need to rename the maps.

Reply:
Oh, I think that's because of another bug I missed. `getClientSessions()` is doing `this._clientMapping.get(externalTaskListId);` but it should be using `_taskListMapping` instead of `_clientMapping`. The usage here (and in that other function) doesn't match the typing as defined by the type aliases. If you update both places, things should work.

(Tangent: this shows how the type aliases, while better than having strings everywhere, aren't infallible 😄).

Reply:
Hmm but getClientSessions isn't used anywhere, and isn't functionally doing anything. So even if the bug is there, it isn't causing anything else to happen.